plutus-core/src/Language/PlutusCore/Flat.hs

```diff
@@ -0,0 +1,349 @@
+that encodes a different number of possibilities. Here is a list of the
+tags and their used/available encoding possibilities.
+
+| Data type       | Function          | Used | Available |
+|-----------------|-------------------|------|-----------|
+| Static builtins | encodeBuiltin     | 22   | 32        |
+| Kinds           | encodeKind        | 2    | 2         |
+| Types           | encodeType        | 7    | 8         |
+| BuiltinNames    | encodeBuiltinName | 2    | 2         |
+| Terms           | encodeTerm        | 10   | 16        |
+
+For format stability we are manually assigning the tag values to the
+constructors (and we do not use a generic algorithm that may change this order).
+
+Flat uses an extra function in its class definition called `size`. Since we want
+to reserve some space for future data constructors and we don't want to have the
+sizes desynchronised from the encoding and decoding functions we have manual
+implementations for them (if they have any constructors reserved for future use).
+-}
+
+{- Note [Don't use catamorphims!]
+We use Codec.Serialise for encoding.  This uses an intermediate type
+`Encoding` to encode things. `Encoding` is a monoid, which allows
+subobjects to be encoded and then efficiently concatenated when a
+larger object is being encoded to Flat.  The monoid structure makes it
+tempting to use catamorphisms to encode things, but this is
+*dangerous*. When a list is encoded to Flat, the start of the list is
+marked by a special token; this is then followed by the encodings of
+the individual list elements, then another token to mark the end of
+the list.  If we use a catamorphism to encode a list `l` it just
+encodes the elements and concatenates the encoded versions together
+without the markers, which is wrong: instead we should call `encode l`
+directly.
+-}
+
+-- See Note [The G, the Tag and the Auto].
+instance Closed uni => Flat (Some (TypeIn uni)) where
+    encode (Some (TypeIn uni)) = encode . map (fromIntegral :: Int -> Word) $ encodeUni uni
+
+    decode = go . decodeUni . map (fromIntegral :: Word -> Int) =<< decode where
+        go Nothing    = fail "Failed to decode a universe"
+        go (Just uni) = pure uni
+
+    -- Encode a view of the universe, not the universe itself.
+    size (Some (TypeIn uni)) acc = size (encodeUni uni) acc
+
+-- See Note [The G, the Tag and the Auto].
+instance (Closed uni, uni `Everywhere` Flat) => Flat (Some (ValueOf uni)) where
+    encode (Some (ValueOf uni x)) = encode (Some $ TypeIn uni) <> bring (Proxy @Flat) uni (encode x)
+
+    decode = go =<< decode where
+        go (Some (TypeIn uni)) = Some . ValueOf uni <$> bring (Proxy @Flat) uni decode
+
+    -- We need to get the flat instance in scope.
+    size (Some (ValueOf uni x)) acc = size (Some $ TypeIn uni) acc
+                                        + bring (Proxy @Flat) uni (size x 0)
+
+encodeBuiltin :: Word8 -> Encoding
+encodeBuiltin = eBits 5
```

Comment:
Worth being explicit: this says we're encoding a number up to 2^5, i.e. 32 as per the table. Maybe even pull the 5 out into a constant, `builtinTagWidth` or something?

Reply:
Same comments for the other one.s

Reply:
I agree.  I looked up `eBits` in http://hackage.haskell.org/package/flat-0.3.4/docs/Data-Flat-Encoder.html, but it doesn't actually explain what it does there.  I don't think it's very complicated, but it would be useful to have a brief explanation here.

Reply:
I have added the suggested constants everywhere.